R/archive.R

```diff
@@ -196,572 +187,755 @@ next_after.Date <- function(x) x + 1L
+#' @param min_time_value Time value specifying the min `time_value` to permit in
+#'   the snapshot. Default is `-Inf`, which effectively means that there is no
+#'   minimum considered.
+#' @param all_versions Boolean; If `all_versions = TRUE`, then the output will be in
+#'   `epi_archive` format, and contain rows in the specified `time_value` range
+#'   having `version <= max_version`. The resulting object will cover a
+#'   potentially narrower `version` and `time_value` range than `x`, depending
+#'   on user-provided arguments. Otherwise, there will be one row in the output
+#'   for the `max_version` of each `time_value`. Default is `FALSE`.
+#' @importFrom data.table between key
+#' @export
+as_of.epi_archive <- function(epi_archive, max_version, min_time_value = -Inf, all_versions = FALSE) {
+  other_keys <- setdiff(
+    key(epi_archive$DT),
+    c("geo_value", "time_value", "version")
+  )
+  if (length(other_keys) == 0) other_keys <- NULL
+
+  # Check a few things on max_version
+  if (!test_set_equal(class(max_version), class(epi_archive$DT$version))) {
+    cli_abort(
+      "`max_version` must have the same classes as `epi_archive$DT$version`."
+    )
+  }
+  if (!test_set_equal(typeof(max_version), typeof(epi_archive$DT$version))) {
+    cli_abort(
+      "`max_version` must have the same types as `epi_archive$DT$version`."
+    )
+  }
+  assert_scalar(max_version, na.ok = FALSE)
+  if (max_version > epi_archive$versions_end) {
+    cli_abort("`max_version` must be at most `epi_archive$versions_end`.")
+  }
+  assert_logical(all_versions, len = 1)
+  if (!is.na(epi_archive$clobberable_versions_start) && max_version >= epi_archive$clobberable_versions_start) {
+    cli_warn(
+      'Getting data as of some recent version which could still be
+      overwritten (under routine circumstances) without assigning a new
+      version number (a.k.a. "clobbered").  Thus, the snapshot that we
+      produce here should not be expected to be reproducible later. See
+      `?epi_archive` for more info and `?epix_as_of` on how to muffle.',
+      class = "epiprocess__snapshot_as_of_clobberable_version"
+    )
+  }
+
+  # Filter by version and return
+  if (all_versions) {
+    # epi_archive is copied into result, so we can modify result directly
+    result <- epix_truncate_versions_after(epi_archive, max_version)
+    result$DT <- result$DT[time_value >= min_time_value, ]
+    return(result)
+  }
+
+  # Make sure to use data.table ways of filtering and selecting
+  as_of_epi_df <- epi_archive$DT[time_value >= min_time_value & version <= max_version, ] %>%
+    unique(
+      by = c("geo_value", "time_value", other_keys),
+      fromLast = TRUE
+    ) %>%
+    tibble::as_tibble() %>%
+    dplyr::select(-"version") %>%
+    as_epi_df(
+      geo_type = epi_archive$geo_type,
+      time_type = epi_archive$time_type,
+      as_of = max_version,
+      additional_metadata = c(epi_archive$additional_metadata,
+        other_keys = other_keys
```

Comment:
```suggestion
        list(other_keys = other_keys)
```
This is hopefully a simple fix for a pre-existing thing that stood out here.  Filed off another one [here](https://github.com/cmu-delphi/epiprocess/issues/437).

`c` should basically never be used on a combination of list/NULL and non-list objects.  Example error:
```r
c(NULL, other_keys = c("sex", "age_group"))
#> other_keys1 other_keys2 
#>       "sex" "age_group" 
```